README.md
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/iota-pico/api-examples-nodejs/master/LICENSE) [![Build Status](https://travis-ci.org/iota-pico/api-examples-nodejs.svg?branch=master)](https://travis-ci.org/iota-pico/api-examples-nodejs) 

# IOTA Pico Framework Examples for NodeJS in TypeScript

Example code for using IOTA Pico Framework with NodeJS written in TypeScript.

# Installation

```shell
npm install @iota-pico/api-examples-nodejs
```

# Running

You can run the examples using the CLI:

```js
node ./bin/ipent
```

## Usage

```shell
Usage: ipent [options] [command]


  Options:

    --neighbors <comma separated uris>       Neighbors to add/remove from your node e.g. udp://1.2.3.4:14600
    --bundles <comma separated hashes>       Hashes to search for in bundles
    --addresses <comma separated hashes>     Hashes to search for or get balances in addresses
    --tags <comma separated hashes>          Hashes to search for in tags
    --approvees <comma separated hashes>     Hashes to search for which confirm specified transaction
    --hashes <comma separated hashes>        Hashes to return the trytes for
    --transactions <comma separated hashes>  Hashes to get inclusion states for
    --tips <comma separated hashes>          Hashes to get inclusion states for
    --threshold <int>                        Confirmation threshold
    --depth <int>                            Number of bundles to go back to determine the transactions for approval
    --trunkTransaction <hash>                Trunk transaction to approve
    --branchTransaction <hash>               Branch transaction to approve
    --minWeightMagnitude <int>               Proof of Work intensity. Minimum value is 18
    --trytes <comma separates trytes>        List of trytes (raw transaction data) to attach/broadcast/store in the tangle
    --tails <comma separated hashes>         Hashes for tails to check consistency
    -V, --version                            output the version number
    -h, --help                               output usage information


  Commands:

    getNodeInfo                         Returns information about your node.
    getNeighbors                        Returns the set of neighbors you are connected with.
    addNeighbors [options]              Add neighbors to your node.
    removeNeighbors [options]           Remove neighbors from your node.
    getTips                             Get the list of tips from the node.
    findTransactions [options]          Find the transactions which match the specified input.
    getTrytes [options]                 Get the trytes for the hashes.
    getInclusionStates [options]        Get the inclusion states of a set of transactions.
    getBalances [options]               Get the balances of a set of transactions.
    getTransactionsToApprove [options]  Tip selection which returns trunkTransaction and branchTransaction.
    attachToTangle [options]            Attaches the specified transactions (trytes) to the Tangle by doing Proof of Work.
    interruptAttachingToTangle          Interrupts and completely aborts the attachToTangle process.
    broadcastTransactions [options]     Broadcast a list of transactions to all neighbors.
    storeTransactions [options]         Store transactions into the local storage.
    getMissingTransactions              Get transactions with missing references.
    checkConsistency [options]          Check the consistency of tail hashes.
    wereAddressesSpentFrom [options]    Have the requested addresses been spent from already.
```

# Configuration

If you want to modify the node that is used to service the examples please modify the config file config/networkConfig.json.

To connect to a node use:

```json
{
    "protocol": "https",
    "host": "nodes.thetangle.org",
    "port": 443
}
```

# Library

The IOTA Pico Framework is a multi-layered set of object oriented JavaScript libraries for use with the IOTA tangle.

If you don't want to use the layered versions of the libraries consider using the  ready bundled versions:

* [@iota-pico/lib-browser](https://github.com/iota-pico/lib-browser)
* [@iota-pico/lib-nodejs](https://github.com/iota-pico/lib-nodejs)

Each layer is fully abstracted allowing you to replace components with your own implementations very easily.

The layered libraries are written in TypeScript so are all strongly typed. The modules are generated as ES6 so you may need to transpile them when including them for use in older JavaScript eco-systems. The code will run without transpilation in all modern browsers and when used by NodeJs.

# Contributing

Contributions are always welcome to the project. Feel free to raise issues, create pull requests or just make suggestions.

# Authors

Come and find us on the IOTA [Discord](https://discord.gg/JJysqe9) development channels

* **Martyn Janes** - *obany* - ([https://github.com/obany](https://github.com/obany))

# Donations

IOTA donations are always welcome :smile:

![pico@tngl-me](https://cdn.tngl.me/tngl-me/pico/qr.svg)

# License

This project is licensed under the MIT License - see the [LICENSE](https://github.com/iota-pico/api-examples-nodejs/blob/master/LICENSE) file for details.


CHANGELOG.md
# Changelog

## v1.0.1

* Refactored NetworkClient to make it more versatile

## v1.0.0

* Final 1.0.0 Release
* Lib files shrunk with different WebPack options and babel removal
* Updated dependencies and README

## v0.9.9

* Updated dependencies and README

## v0.9.8

* Updated dependencies

## v0.9.7

* Reduced umd module size with externals

## v0.9.6

* Dependency Updates

## v0.9.5

* GitHub and Npm Version alignment

## v0.9.4

* Fixed repo links in readme and coverage config

## v0.9.3

* Repo rename

## v0.9.2

* Added UMD version of library in pkg folder

## v0.9.1

* Fix Release


src/index.ts
import chalk from "chalk";
import commander from "commander";

/**
 * Main CLI Interface
 */
export async function run(): Promise<void> {
    const packageJson = require("../package.json");

    const oldConsoleError = console.error;
    console.error = (error: any) => {
        oldConsoleError(error ? chalk.red(error) : "");
    };

    console.log(packageJson.description);
    console.log("".padStart(packageJson.description.length, "="));

    commander
        .option("--neighbors <comma separated uris>", "Neighbors to add/remove from your node e.g. udp://1.2.3.4:14600", (val: string) => val.split(","))
        .option("--bundles <comma separated hashes>", "Hashes to search for in bundles", (val: string) => val.split(","))
        .option("--addresses <comma separated hashes>", "Hashes to search for or get balances in addresses", (val: string) => val.split(","))
        .option("--tags <comma separated hashes>", "Hashes to search for in tags", (val: string) => val.split(","))
        .option("--approvees <comma separated hashes>", "Hashes to search for which confirm specified transaction", (val: string) => val.split(","))
        .option("--hashes <comma separated hashes>", "Hashes to return the trytes for", (val: string) => val.split(","))
        .option("--transactions <comma separated hashes>", "Hashes to get inclusion states for", (val: string) => val.split(","))
        .option("--tips <comma separated hashes>", "Hashes to get inclusion states for", (val: string) => val.split(","))
        .option("--threshold <int>", "Confirmation threshold", parseInt)
        .option("--depth <int>", "Number of bundles to go back to determine the transactions for approval", parseInt)
        .option("--trunkTransaction <hash>", "Trunk transaction to approve")
        .option("--branchTransaction <hash>", "Branch transaction to approve")
        .option("--minWeightMagnitude <int>", "Proof of Work intensity. Minimum value is 18", parseInt)
        .option("--trytes <comma separates trytes>", "List of trytes (raw transaction data) to attach/broadcast/store in the tangle", (val: string) => val.split(","))
        .option("--tails <comma separated hashes>", "Hashes for tails to check consistency", (val: string) => val.split(","))
        .version(packageJson.version);

    commander
        .command("getNodeInfo")
        .description("Returns information about your node.")
        .action(async () => {
            await runExample("getNodeInfo");
        });

    commander
        .command("getNeighbors")
        .description("Returns the set of neighbors you are connected with.")
        .action(async () => {
            await runExample("getNeighbors");
        });

    commander
        .command("addNeighbors")
        .option("--neighbors <uri>")
        .description("Add neighbors to your node.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.neighbors) {
                console.error("ERROR: neighbors option is required");
                return;
            }
            await runExample("addNeighbors", cmd.parent.neighbors);
        });

    commander
        .command("removeNeighbors")
        .option("--neighbors <uri>")
        .description("Remove neighbors from your node.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.neighbors) {
                console.error("ERROR: neighbor option is required");
                return;
            }
            await runExample("removeNeighbors", cmd.parent.neighbors);
        });

    commander
        .command("getTips")
        .description("Get the list of tips from the node.")
        .action(async () => {
            await runExample("getTips");
        });

    commander
        .command("findTransactions")
        .option("--bundles <comma separated hashes>")
        .option("--addresses <comma separated hashes>")
        .option("--tags <comma separated hashes>")
        .option("--approvees <comma separated hashes>")
        .description("Find the transactions which match the specified input.")
        .action(async (cmd) => {
            if (!cmd.parent || !(cmd.parent.bundles || cmd.parent.addresses || cmd.parent.tags || cmd.parent.approvees)) {
                console.error("ERROR: bundles/addresses/tags/approvees option is required");
                return;
            }
            await runExample("findTransactions", cmd.parent.bundles, cmd.parent.addresses, cmd.parent.tags, cmd.parent.approvees);
        });

    commander
        .command("getTrytes")
        .option("--hashes <comma separated hashes>")
        .description("Get the trytes for the hashes.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.hashes) {
                console.error("ERROR: hashes option is required");
                return;
            }
            await runExample("getTrytes", cmd.parent.hashes);
        });

    commander
        .command("getInclusionStates")
        .option("--transactions <comma separated hashes>")
        .option("--tips <comma separated hashes>")
        .description("Get the inclusion states of a set of transactions.")
        .action(async (cmd) => {
            if (!cmd.parent || !(cmd.parent.transactions || cmd.parent.tips)) {
                console.error("ERROR: transactions/tips option is required");
                return;
            }
            await runExample("getInclusionStates", cmd.parent.transactions || [], cmd.parent.tips || []);
        });

    commander
        .command("getBalances")
        .option("--addresses <comma separated hashes>")
        .option("--threshold <int>")
        .description("Get the balances of a set of transactions.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.addresses) {
                console.error("ERROR: addresses option is required");
                return;
            }
            await runExample("getBalances", cmd.parent.addresses, cmd.parent.threshold || 100);
        });

    commander
        .command("getTransactionsToApprove")
        .option("--depth <int>")
        .description("Tip selection which returns trunkTransaction and branchTransaction.")
        .action(async (cmd) => {
            if (!cmd.parent || cmd.parent.depth === undefined) {
                console.error("ERROR: depth option is required");
                return;
            }
            await runExample("getTransactionsToApprove", cmd.parent.depth);
        });

    commander
        .command("attachToTangle")
        .option("--trunkTransaction <hash>")
        .option("--branchTransaction <hash>")
        .option("--minWeightMagnitude <number>")
        .option("--trytes <comma separates trytes>")
        .description("Attaches the specified transactions (trytes) to the Tangle by doing Proof of Work.")
        .action(async (cmd) => {
            if (!cmd.parent || !(cmd.parent.trunkTransaction || cmd.parent.branchTransaction || cmd.parent.minWeightMagnitude === undefined || cmd.parent.trytes)) {
                console.error("ERROR: trunkTransaction/branchTransaction/minWeightMagnitude/trytes options are required");
                return;
            }
            await runExample("attachToTangle", cmd.parent.trunkTransaction, cmd.parent.branchTransaction, cmd.parent.minWeightMagnitude, cmd.parent.trytes);
        });

    commander
        .command("interruptAttachingToTangle")
        .description("Interrupts and completely aborts the attachToTangle process.")
        .action(async (cmd) => {
            await runExample("interruptAttachingToTangle");
        });

    commander
        .command("broadcastTransactions")
        .option("--trytes <comma separates trytes>")
        .description("Broadcast a list of transactions to all neighbors.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.trytes) {
                console.error("ERROR: trytes option is required");
                return;
            }
            await runExample("broadcastTransactions", cmd.parent.trytes);
        });

    commander
        .command("storeTransactions")
        .option("--trytes <comma separates trytes>")
        .description("Store transactions into the local storage.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.trytes) {
                console.error("ERROR: trytes option is required");
                return;
            }
            await runExample("storeTransactions", cmd.parent.trytes);
        });

    commander
        .command("getMissingTransactions")
        .description("Get transactions with missing references.")
        .action(async (cmd) => {
            await runExample("getMissingTransactions");
        });

    commander
        .command("checkConsistency")
        .option("--tails <comma separated hashes>")
        .description("Check the consistency of tail hashes.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.tails) {
                console.error("ERROR: tails option is required");
                return;
            }
            await runExample("checkConsistency", cmd.parent.tails);
        });

    commander
        .command("wereAddressesSpentFrom")
        .option("--addresses <comma separated hashes>")
        .description("Have the requested addresses been spent from already.")
        .action(async (cmd) => {
            if (!cmd.parent || !cmd.parent.addresses) {
                console.error("ERROR: addresses option is required");
                return;
            }
            await runExample("wereAddressesSpentFrom", cmd.parent.addresses);
        });

    commander
        .command("*", undefined, { noHelp: true })
        .action(() => {
            console.error(`Unknown Command: ${commander.args.join(" ")}`);
            commander.help();
        });

    commander.parse(process.argv);

    if (!process.argv.slice(2).length) {
        commander.help();
    }
}

async function runExample(api: string, ...args: any[]): Promise<void> {
    try {
        const oldConsoleInfo = console.info;
        console.info = (info: any) => {
            oldConsoleInfo(info ? chalk.cyan(info) : "");
        };

        const oldConsoleLog = console.log;
        console.log = (log: any) => {
            oldConsoleLog(log ? chalk.green(log) : "");
        };

        const module = await require(`./api/${api}Example`);
        const exampleMethod = module[`${api}Example`];
        exampleMethod(...args);
    } catch (err) {
        console.log(chalk.red(`Unable to load example ${api}.`));
    }
}
